File: 백준/Gold/13265. 색칠하기/색칠하기.java
import java.util.*;
import java.io.*;

public class Main {

  static ArrayList[] G;

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    int T = Integer.parseInt(br.readLine());
    while (T-- > 0) {
      StringTokenizer st = new StringTokenizer(br.readLine());
      int n = Integer.parseInt(st.nextToken());
      int m = Integer.parseInt(st.nextToken());
      G = new ArrayList[n + 1];
      for (int i = 0; i <= n; i++) {
        G[i] = new ArrayList<Integer>();
      }
      while (m-- > 0) {
        st = new StringTokenizer(br.readLine());
        int x = Integer.parseInt(st.nextToken());
        int y = Integer.parseInt(st.nextToken());
        G[x].add(y);
        G[y].add(x);
      }
      int[] visited = new int[n + 1];
      String ans = "possible";
      for (int i = 1; i <= n; i++) {
        if (visited[i] == 0) {
          visited[i] = 1;
          if (!dfs(i, visited)) {
            ans = "impossible";
            break;
          }
        }
      }
      sb.append(ans).append("\n");
    }
    System.out.println(sb);
  }

  static boolean dfs(int cur, int[] visited) {
    boolean res = true;
    ArrayList<Integer> nextNodes = G[cur];
    for (Integer nxt : nextNodes) {
      if (visited[nxt] == 0) {
        visited[nxt] = -visited[cur];
        res = dfs(nxt, visited);
      } else if (visited[nxt] == visited[cur]) {
        return false;
      }
    }
    return res;
  }

}

File: 백준/Gold/13265. 색칠하기/색칠하기.py
from collections import deque
import sys
input = sys.stdin.readline

def bfs(start, G, visited):
    q = deque([start])
    visited[start] = 1
    while q:
        cur = q.popleft()
        for nxt in G[cur]:
            if not visited[nxt]:
                q.append(nxt)
                visited[nxt] = -visited[cur]
            elif visited[nxt] == visited[cur]:
                return False
    return True

def solve():
    T = int(input())
    for _ in range(T):
        n, m = map(int, input().split())
        G = [[] for _ in range(n + 1)]
        for _ in range(m):
            x, y = map(int, input().split())
            G[x].append(y)
            G[y].append(x)
        visited = [0] * (n + 1)
        for i in range(1, n + 1):
            if not visited[i]:
                if bfs(i, G, visited) == False:
                    print("impossible")
                    break
        else:
            print("possible")

solve()